### ISSUE TITLE
Conditional formatting for Nullable values

### ISSUE THREAD

**Issue by thexray** [scottrippey/SmartFormat.NET#39]
I do not understand from the [Conditional Formatting Syntax](https://github.com/scottrippey/SmartFormat.NET/wiki/Conditional-Formatting-Syntax) chapter  how to test for null values of numeric properties.

This is an example that doesn't produce expected result:

```c#
using System;
using SmartFormat;

public class Point
{
	public double? Lat { get; set; }
	public double? Lon { get; set; }
}

public class Program
{
	public static void Main()
	{
		var point = new Point {Lat = 23.45};
		Console.WriteLine(Smart.Format("Latitude: {Lat:{}|N/A}, longitude: {Lon:{}|N/A}", point));
		
		// Output is:
		// Latitude: N/A, longitude: N/A
		
		// Output I want is:
		// Latitude: 23.45, longitude: N/A
	}
}
```
You can run it in this [fiddle](https://dotnetfiddle.net/ucEwzA).

**Comment by scottrippey** [comment]
Well, there currently is no special support for `Nullable<T>` in the `ConditionalFormatter`, which means it should default to the "value|null" behavior, as you expected.
And, indeed, the logic at [L155 of ConditionalFormatter.cs](https://github.com/scottrippey/SmartFormat.NET/blob/master/src/SmartFormat/Extensions/ConditionalFormatter.cs#L155-L166)
compares the value to `null`.

So, I'd qualify this as a bug.

**Comment by scottrippey** [comment]
Perhaps this is caused by "boxing" the Nullable in an object?  I thought that Nullables still work, even when boxed.

**Comment by SeriousM** [comment]
Related: #39

**Comment by thexray** [comment]
As a quick workaround I've changed [L91-98 of ConditionalForamtter.cs](https://github.com/scottrippey/SmartFormat.NET/blob/master/src/SmartFormat/Extensions/ConditionalFormatter.cs#L91-L98) to
```c#
paramIndex = 0;
```

**Comment by scottrippey** [comment]
@thexray That's unexpected ... if that worked, then it means "currentIsNumber" must be true, which means `current is double` returns True?  Do Nullables work this way?

**Comment by scottrippey** [comment]
Ok, I figured some stuff out...
As seen in [this fiddle](https://dotnetfiddle.net/biH7Mg), `(decimal?)null is decimal` is `False`, but `(decimal?)5 is decimal` is `True`!  This is quite surprising, but I guess it makes sense.  I'm guessing that the Nullable type has an implicit type coverter to Object that causes this behavior.

**Comment by scottrippey** [comment]
So, this still provides an interesting challenge.  When the Nullable is casted to an Object, it actually "unwraps" itself into either null, or the native number type.  This means that "null" values will use the "object" logic, but non-null values will use the Number logic.
I think we need to explicitly check for the "Nullable" data type, and force it to use the "object" logic.
Is there an easy way to check if a type is nullable?  (Something better than a bunch of `current is decimal?` statements?)

**Comment by SeriousM** [comment]
Yes it is!

` if (type.IsGenericType && type.GetGenericTypeDefinition() ==
typeof(Nullable<>)) {…} `

Source: http://msdn.microsoft.com/en-us/library/ms366789.aspx

**Comment by scottrippey** [comment]
Ok, thanks for the link.  According to that link, it's NOT possible:
"
Calling GetType on a Nullable type causes a boxing operation to be performed when the type is implicitly converted to Object. Therefore GetType always returns a Type object that represents the underlying type, not the Nullable type.
"

To put it in my own words: when any `Nullable` is cast to `object`, it loses its "nullability".  It either gets cast to `null`, or cast to the underlying data type.  
So within the Conditional Formatter code, it's actually impossible to tell if `current` was nullable, because it was converted to `object` and lost its original type.

This actually might be extremely difficult to solve.

**Comment by SeriousM** [comment]
Well, it's not possible in our case :)

I didn't had a look at our code...

Can we save the original value anyhow? Or store the nullability info
separately as meta data?

**Comment by scottrippey** [comment]
Well, I think the only possibility is for the Reflection plugin to detect if it's Nullable.  
Then it'd have to store that as a flag somewhere.  I don't think any of this is trivial.

**Comment by scottrippey** [comment]
I've got a solution in the works.  
I've been working on a "named formatters" feature that allows you to choose which formatter you want to use, in addition to passing in options.  
This feature will allow you to specifically check for "null" values, and won't use "index" logic if it's a number.
This feature is slated for v1.3.

**Comment by scottrippey** [comment]
Alright, this feature has been implemented, and v1.4 is released.  
Basically, there's a new "choose" binding that lets you specifically target null values.  

This ought to work:
    Smart.Format("Latitude: {Lat:choose(null):N/A|{}}, longitude: {Lon:choose(null):N/A|{}}", point);
